### DISCUSSION TITLE
Pyright handling items method of TypedDict

### DISCUSSION THREAD

**Discussion by mheguy** [microsoft/pyright#9162] [Q&A] [answered]
Heya, the behaviour of the following is unexpected:

```python
from typing import TypedDict

class TestClass(TypedDict):
    input: str

for _, bloop in TestClass(input="input").items():
    _.lower()
    bloop.lower() # error: Cannot access attribute "lower" for class "object" - Attribute "lower" is unknown (reportAttributeAccessIssue)
```

This doesn't seem like a bug in Pyright, it's just doing what the type hints say.
The typeshed says: `def items(self) -> dict_items[str, object]: ...`

Is there a (reasonable) way to work around this?

**Comment by erictraut** [comment, marked as answer]
Yeah, pyright is working correctly here. The way `TypedDict` is defined in the typing spec, it is a structural type, and a value that is compatible with a TypedDict can contain additional items beyond what that TypedDict defines. That means when iterating over all items, a type checker needs to assume there are potentially additional items present whose values are `object` type.

If you know that there are no additional items present, you can add an `assert isinstance(bloop, str)` statement in your loop.

You may also be interested in draft [PEP 728](https://peps.python.org/pep-0728/), which proposes to add support for "closed" TypedDict definitions. This PEP is in flux, but pyright has provisional support for the latest draft if you would like to play with it. You need to set `enableExperimentalFeatures` to true.
